dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/token/TokenRequestor.java

```diff
@@ -0,0 +1,204 @@
+        if (customParams == null || customParams.isEmpty()) {
+            return;
+        }
+
+        for (Entry<String, String> param : customParams.entrySet()) {
+            if (param.getKey() != null && param.getValue() != null) {
+                params.add(new BasicNameValuePair(param.getKey(), param.getValue()));
+            }
+        }
+    }
+
+    public TokenResponse requestTokens() throws Exception {
+        Map<String, Object> tokenEndpointResponse = postToTokenEndpoint();
+        String tokenEndpointEntity = oidcClientHttpUtil.extractEntityFromTokenResponse(tokenEndpointResponse);
+        Map<String, String> tokens = getTokensFromJsonString(tokenEndpointEntity);
+        return new TokenResponse(tokens.get(TokenConstants.ID_TOKEN), tokens.get(TokenConstants.ACCESS_TOKEN), tokens.get(TokenConstants.REFRESH_TOKEN));
+    }
+
+    private Map<String, Object> postToTokenEndpoint() throws Exception {
+        return oidcClientHttpUtil.postToEndpoint(
+                                                 tokenEndpoint,
+                                                 params,
+                                                 clientId,
+                                                 clientSecret,
+                                                 null,
+                                                 sslSocketFactory,
+                                                 commonHeaders,
+                                                 isHostnameVerification,
+                                                 authMethod,
+                                                 useSystemPropertiesForHttpClientConnections);
+    }
+
+    private Map<String, String> getTokensFromJsonString(String jsonString) throws Exception {
+        Map<String, String> tokens = new HashMap<>();
+        List<String> tokenTypes = Arrays.asList(TokenConstants.TOKEN_TYPES);
+
+        JSONObject json = JSONObject.parse(jsonString);
+
+        @SuppressWarnings("unchecked")
+        Iterator<String> iterator = json.keySet().iterator();
+        while (iterator.hasNext()) {
+            String key = iterator.next();
+            Object value = json.get(key);
+            if (value instanceof String && tokenTypes.contains(key)) {
+                tokens.put(key, value.toString());
+            }
+        }
+
+        return tokens;
+    }
+
+    public static class Builder {
+
+        private final String grantType = TokenConstants.AUTHORIZATION_CODE;
+
+        private final String tokenEndpoint;
+        private final String clientId;
+        @Sensitive
+        private final String clientSecret;
+        private final String redirectUri;
+        private final String code;
+
+        private SSLSocketFactory sslSocketFactory = null;
+        private boolean isHostnameVerification = false;
+        private String authMethod = TokenConstants.METHOD_BASIC;
+        private String resources = "";
```

Comment:
Should this default to null instead so we're not sending a `resource=` parameter in the token request?

Reply:
yeah you're right, thanks for catching this 🙂